manager/integration/tests/test_csi_snapshotter.py

```diff
@@ -674,23 +732,213 @@ def test_csi_snapshot_with_snap_param(client, volume_name): # NOQA
+                            ready_to_use=True)
+
+    # Step 1 Test create new volume from CSI snapshot
+    # Source volume is attached && Longhorn snapshot exist
+    pvc['spec']['storageClassName'] = 'longhorn'
+    pvc['spec']['dataSource'] = {
+        'name': csivolsnap["metadata"]["name"],
+        'kind': 'VolumeSnapshot',
+        'apiGroup': 'snapshot.storage.k8s.io'
+    }
+    pvc['spec']['resources']['requests']['storage'] = str(500 * Mi)
+    create_pvc(pvc)
+
+    pv_name = wait_and_get_pv_for_pvc(core_api,
+                                      pvc['metadata']['name']).metadata.name
+    new_deployment_name = pv_name + "-dep"
+    new_deployment = make_deployment_with_pvc(new_deployment_name,
+                                              pvc['metadata']['name'])
+    new_deployment["spec"]["replicas"] = 1
+    apps_api.create_namespaced_deployment(body=new_deployment,
+                                          namespace='default')
+
+    common.wait_for_volume_status(client, pv_name,
+                                  common.VOLUME_FIELD_STATE,
+                                  common.VOLUME_STATE_ATTACHED)
+    data_path = "/data/test"
+    pod = common.wait_and_get_any_deployment_pod(core_api, new_deployment_name)
+    created_md5sum = get_pod_data_md5sum(core_api, pod.metadata.name,
+                                         data_path)
+
+    assert expected_md5sum == created_md5sum
+
+    # Source volume is detached
+    deployment["spec"]["replicas"] = 0
+    apps_api.patch_namespaced_deployment(body=deployment,
+                                         namespace='default',
+                                         name=deployment_name)
+    common.wait_for_volume_status(client, vol.name,
+                                  common.VOLUME_FIELD_STATE,
+                                  common.VOLUME_STATE_DETACHED)
+
+    new_pvc1 = pvc
+    new_pvc1['metadata']['name'] = pvc['metadata']['name'] + "new-pvc1"
+    create_pvc(new_pvc1)
+    check_pvc_in_specific_status(core_api,
+                                 new_pvc1['metadata']['name'], "Pending")
+
+    # Source volume is attached && Longhorn snapshot doesn’t exist
+    deployment["spec"]["replicas"] = 1
+    apps_api.patch_namespaced_deployment(body=deployment,
+                                         namespace='default',
+                                         name=deployment_name)
+    common.wait_for_volume_status(client, vol.name,
+                                  common.VOLUME_FIELD_STATE,
+                                  common.VOLUME_STATE_ATTACHED)
+
+    vol = client.by_id_volume(vol.name)
+    # create new snapshot aovid only exist 1 snpashot and can not deleted
+    vol.snapshotCreate()
+    snapshots = vol.snapshotList()
+    for item in snapshots:
+        if "snap" in item.name:
+            vol.snapshotDelete(name=item.name)
+    vol.snapshotPurge()
+
+    new_pvc2 = pvc
+    new_pvc2['metadata']['name'] = pvc['metadata']['name'] + "new-pvc2"
```

Comment:
Can we make the logic here a bit clearer for the reader by
```suggestion
    # Find the VolumeSnapshotContent of the volumesnapshot_name.
    # Find the Longhorn snapshot name inside the field VolumeSnapshotContent.snapshotHandle, let's call it snap_name
    snapshots = vol.snapshotList()
    for item in snapshots:
        if snap_name == item.name:
            vol.snapshotDelete(name=snap_name)
    vol.snapshotPurge()
    # wait for snap_name to be removed from the volume

    new_pvc2 = pvc
    new_pvc2['metadata']['name'] = pvc['metadata']['name'] + "new-pvc2"
```

Reply:
HI @PhanLe1010 
I modified coed for review, thanks